### PR TITLE
Update shared gateway to support IPv6/Dualstack

### PR DESCRIPTION
**Pull request by Billy99**
Changes to shared gateway mode to support IPv6 Only and Dual-stack. Addresses issue #1141.

Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by Billy99** [comment]
This PR depends on PR #1343, which contains other IPv6 only changes.

Even with the additional IPv6 changes, this PR is still failing. I believe there are issues with the IPv6 NAT.

**Comment by danwinship** [comment]
I don't think it really makes sense to add IPv6-only support first at this point, since we'll just have to change all the same functions again to make it dual-stack. Just go right to dual-stack.

**Comment by Billy99** [comment]
I was originally part of PR #1343, which was getting to big, so I broke it off. The only reason to separate from dual-stack is that IPv6 only isn't working in shared gateway mode and is probably a big enough fix to be stand alone. But agree that the code will need to be reworked again for dual-stack, so I'll head in that direction.

File: go-controller/pkg/node/gateway_init.go [hunk 1]
```diff
@@ -105,15 +112,23 @@ func (n *OvnNode) initGateway(subnets []*net.IPNet, nodeAnnotator kube.Annotator
 		}
 	}
 
-	gatewayNextHop := net.ParseIP(config.Gateway.NextHop)
+	var gatewayNextHops []net.IP
+	// TODO:
+	//   Should config.Gateway.NextHop be a slice of nexthops?
+	//   If so, if dual-stack, but only one NextHop provided, what then?
```
Comment (danwinship) [inline go-controller/pkg/node/gateway_init.go:118]: We aren't going to be using this functionality anyway. We can leave it single-stack for now. In the dual-stack case, just always use the default next hop for the IP family that doesn't have an explicitly-specified next hop

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 2]
```diff
@@ -465,16 +471,17 @@ func (n *OvnNode) initSharedGateway(subnets []*net.IPNet, gwNextHop net.IP, gwIn
...
-	ipAddress, err := getIPv4Address(gwIntf)
+	ips, err := getNetworkInterfaceIPAddresses(gwIntf)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get interface details for %s (%v)",
 			gwIntf, err)
 	}
-	if ipAddress == nil {
-		return nil, fmt.Errorf("%s does not have a ipv4 address", gwIntf)
+	if len(ips) == 0 {
+		return nil, fmt.Errorf("%s does not have an IP address", gwIntf)
```
Comment (danwinship) [inline go-controller/pkg/node/gateway_shared_intf.go:482]: Need to make sure we have an IP address corresponding to each family in `subnets`

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 3]
```diff
@@ -496,8 +503,8 @@ func (n *OvnNode) initSharedGateway(subnets []*net.IPNet, gwNextHop net.IP, gwIn
 		ChassisID:      chassisID,
 		InterfaceID:    ifaceID,
 		MACAddress:     macAddress,
-		IPAddresses:    []*net.IPNet{ipAddress},
-		NextHops:       []net.IP{gwNextHop},
+		IPAddresses:    ips,
```
Comment (danwinship) [inline go-controller/pkg/node/gateway_shared_intf.go:506]: `IPAddresses` in the `L3GatewayConfig` is not just the raw list of all available node IPs; if this is a single-stack cluster (ie, `len(subnets) == 1`) then `IPAddresses` needs to have exactly 1 address, of the correct family. If this is a dual-stack cluster, it needs to have exactly 2 addresses, one of each family. (They don't necessarily need to be in the same order as `subnets`.)

File: go-controller/pkg/node/helper_linux.go [hunk 4]
```diff
@@ -7,35 +7,74 @@ import (
...
-		if route.Dst == nil && route.Gw != nil && route.LinkIndex > 0 {
-			intfLink, err := netlink.LinkByIndex(route.LinkIndex)
-			if err != nil {
-				continue
-			}
-			intfName := intfLink.Attrs().Name
-			if intfName != "" {
-				return intfName, route.Gw, nil
+	for _, family := range familyList {
+		routes, err := netlink.RouteList(nil, family)
```
Comment (danwinship) [inline go-controller/pkg/node/helper_linux.go:38]: You can use `syscall.AF_UNSPEC` to get back the IPv4 and IPv6 routes together... that might be easier?

File: go-controller/pkg/node/helper_linux.go [hunk 5]
```diff
@@ -7,35 +7,74 @@ import (
...
-// It also returns the default gateway itself.
-func getDefaultGatewayInterfaceDetails() (string, net.IP, error) {
-	routes, err := netlink.RouteList(nil, syscall.AF_INET)
-	if err != nil {
-		return "", nil, fmt.Errorf("failed to get routing table in node")
+// It also returns the default gateways themselves.
+func getDefaultGatewayInterfaceDetails() (string, []net.IP, error) {
+	var familyList []int
+	var intfName string
+	var gatewayIps []net.IP
```
Comment (danwinship) [inline go-controller/pkg/node/helper_linux.go:24]: `gatewayIPs` (capital "P")

File: go-controller/pkg/ovn/gateway_init.go [hunk 6]
```diff
@@ -316,13 +317,38 @@ func addDistributedGWPort() error {
...
+	var natSubnetNextHop string
+	if config.IPv4Mode {
+		dgpIP := net.ParseIP(util.V4NodeLocalDistributedGwPortIP)
+		cidrs = append(cidrs,
+			fmt.Sprintf("%s/%d", dgpIP.String(), util.V4NodeLocalNatSubnetPrefix))
+		macStr = util.IPAddrToHWAddr(dgpIP).String()
+		natSubnetNextHop = util.V4NodeLocalNatSubnetNextHop
+	}
+	if config.IPv6Mode {
+		dgpIP := net.ParseIP(util.V6NodeLocalDistributedGwPortIP)
```
Comment (danwinship) [inline go-controller/pkg/ovn/gateway_init.go:331]: There needs to be some sort of loop here for dual-stack, to configure both the ipv4 and ipv6 versions

File: go-controller/pkg/ovn/loadbalancer.go [hunk 7]
```diff
@@ -187,15 +188,21 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb string, sourceIP string, s
...
+	//aclName = strings.ReplaceAll(aclName, ":", "\\:")
+	//aclName = aclName[:63]
+	if len(aclName) > 63 {
+		klog.Infof("BILLY: Limiting ACL Name from %s to %.63s to keep under 63 characters", aclName, aclName)
+	}
 	// If ovn-k8s was restarted, we lost the cache, and an ACL may already exist in OVN. In that case we need to check
 	// using ACL name
 	aclUUID, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading", "--columns=_uuid", "find", "acl",
-		fmt.Sprintf("name=%s", strings.ReplaceAll(aclName, ":", "\\:")))
+		fmt.Sprintf("name=%.63s", strings.ReplaceAll(aclName, ":", "\\:")))
```
Comment (danwinship) [inline go-controller/pkg/ovn/loadbalancer.go:201]: We can't just do this; the end of the `aclName` is important. You might create duplicate names by truncating.

(This seems like an unrelated bugfix anyway? File a new issue/PR for this?)

File: go-controller/pkg/util/net.go [hunk 8]
```diff
@@ -172,6 +173,13 @@ func IPAddrToHWAddr(ip net.IP) net.HardwareAddr {
 		return net.HardwareAddr{0x0A, 0x58, ip4[0], ip4[1], ip4[2], ip4[3]}
 	}
 
+	if (ip[0] == 0xFD && ip[1] == 0x98) || (ip[0] == 0xFD && ip[1] == 0x00) {
+		klog.Errorf("BILLY: IPAddrToHWAddr() MAC=%02x:%02x:%02x:%02x:%02x:%02x IP=%s", 0x0A, 0x58, ip[0], ip[1], ip[7], ip[15], ip.String())
+		// IPv6 - use the first two and last two bytes.
+		return net.HardwareAddr{0x0A, 0x58, ip[0], ip[1], ip[7], ip[15]}
```
Comment (danwinship) [inline go-controller/pkg/util/net.go:179]: we probably need to hash the IP or something

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 2 above)
Comment (Billy99) [reply go-controller/pkg/node/gateway_shared_intf.go:482]: Moved error handling to within getNetworkInterfaceIPAddresses().
Done.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 3 above)
Comment (Billy99) [reply go-controller/pkg/node/gateway_shared_intf.go:506]: Updated getNetworkInterfaceIPAddresses() to only add one address from each family, if enabled. I add the first address found in a family and skip any others found (not sure there would ever be more than one per family, but now it is enforced).
Done.

File: go-controller/pkg/node/gateway_init.go [hunk 9]
```diff
@@ -82,15 +82,22 @@ func getIPv4Address(iface string) (*net.IPNet, error) {
...
+					ips = append(ips, ip)
+				}
+			} else {
+				if config.IPv4Mode {
+					ips = append(ips, ip)
+				}
 			}
 		}
 	}
-	return nil, nil
```
[the remark this answers is not in the document]
Comment (danwinship) [reply go-controller/pkg/node/gateway_init.go:93]: we generally don't assume anything about the ordering of IP arrays

[the remark this answers is not in the document]
Comment (gvbalaji) [reply go-controller/pkg/node/gateway_init.go:93]: OK. This is used in `initSharedGateway` method. It passes an array of ip addreses and array of gw addresses as parameters to  `util.SetL3GatewayConfig`
I guess may be they need to be in the same order?  If so either this need to be handled on the client side or the function return value.

**Comment by gvbalaji** [comment]
thanks @Billy99 . Replied to the ordering of IP addresses as the array of ips and nexthops are passed into util.SetL3GatewayConfig method.  Not sure if that method requires right ordering of ips. If so we need to handle that.

File: contrib/kind.sh [hunk 10]
```diff
@@ -34,6 +34,37 @@ delete()
   kind delete cluster --name ${KIND_CLUSTER_NAME:-ovn}
 }
 
+update_kubeconfig()
+{
+  # For IPv6 NO-HA Shared-Gateway Mode, it is a pure IPv6 network and no IPv4 connectivity.
+  # So need to updae the KubeConfig with an IPv6 address. In HA mode, there is a load-balancer
```
Comment (danwinship) [inline contrib/kind.sh:40]: "update"

File: contrib/kind.sh [hunk 11]
```diff
@@ -34,6 +34,37 @@ delete()
...
+  # in front which responds to the IPv4 requests.
+  if [ "$IP_FAMILY" == "ipv6" ] && [ "$OVN_GATEWAY_MODE" == "shared" ] && [ "$KIND_HA" == false ]; then
+    # Determine if API_URL has an IP Address or Container Name
+    SERVER=`echo "$API_URL" | sed -e "s@^https://@@" -e "s@:6443@@"`
+    if [[ "$SERVER" == *":"* ]]; then
```
Comment (danwinship) [inline contrib/kind.sh:45]: `API_URL` will contain brackets if and only if it contains an IPv6 address, so you could just do

    if [[ "$API_URL" == *"["* ]]; then

File: contrib/kind.sh [hunk 12]
```diff
@@ -34,6 +34,37 @@ delete()
...
+      # IP Address, so leave as is
+      SERVER=$API_URL
+    else
+      # Remove - Debugging
+      docker container inspect ${SERVER}
```
Comment (danwinship) [inline contrib/kind.sh:50]: do you still need that?

File: contrib/kind.sh [hunk 13]
```diff
@@ -34,6 +34,37 @@ delete()
   kind delete cluster --name ${KIND_CLUSTER_NAME:-ovn}
 }
 
+update_kubeconfig()
+{
+  # For IPv6 NO-HA Shared-Gateway Mode, it is a pure IPv6 network and no IPv4 connectivity.
+  # So need to updae the KubeConfig with an IPv6 address. In HA mode, there is a load-balancer
+  # in front which responds to the IPv4 requests.
```
Comment (danwinship) [inline contrib/kind.sh:41]: Is IPv6 NO-HA Shared-Gateway the only case where API_URL ends up with a container name in it rather than an IP address? If not, why is it that the container name works in other cases but needs to be rewritten here? If we can resolve the container name to an IP in other cases, why can't we here? (Does the container name resolve to an IPv4 address? Can we fix that?)

File: contrib/kind.sh [hunk 14]
```diff
@@ -34,6 +34,37 @@ delete()
...
+      else
+        GBLIPV6ADDR=`docker container inspect -f '{{ .NetworkSettings.Networks.bridge.GlobalIPv6Address }}' ${SERVER}`
+        if [ "$GBLIPV6ADDR" != "<no value>" ]; then
+          SERVER="https://[$GBLIPV6ADDR]:6443"
+        else
+          SERVER=$API_URL
+        fi
+      fi
+    fi
+    sed -i 's@\(server:\)\(.*\)@\1 '"${SERVER}"'@' ${KUBECONFIG}
```
Comment (danwinship) [inline contrib/kind.sh:65]: do we need to update `API_URL` for the other parts of kind setup that are using it?

File: dist/yaml/.gitignore [hunk 15]
```diff
@@ -1,3 +1,5 @@
+k8s.ovn.org_egressfirewalls.yaml
+k8s.ovn.org_egressips.yaml
```
Comment (danwinship) [inline dist/yaml/.gitignore:2]: please file this as a separate PR

File: go-controller/pkg/node/gateway_init.go [hunk 16]
```diff
@@ -82,15 +82,35 @@ func getIPv4Address(iface string) (*net.IPNet, error) {
...
+	var foundIPv4 bool
+	var foundIPv6 bool
 	for _, addr := range addrs {
 		switch ip := addr.(type) {
 		case *net.IPNet:
-			if !utilnet.IsIPv6CIDR(ip) {
-				return ip, nil
+			if utilnet.IsIPv6CIDR(ip) {
+				if config.IPv6Mode {
+					if !foundIPv6 {
```
Comment (danwinship) [inline go-controller/pkg/node/gateway_init.go:93]: could maybe combine these with `&&`...

File: go-controller/pkg/node/gateway_init.go [hunk 17]
```diff
@@ -105,15 +125,45 @@ func (n *OvnNode) initGateway(subnets []*net.IPNet, nodeAnnotator kube.Annotator
...
+					gatewayNextHops = append(gatewayNextHops, defaultGatewayNextHop)
+					needIPv4NextHop = false
+				} else if needIPv6NextHop && utilnet.IsIPv6(defaultGatewayNextHop) {
+					gatewayNextHops = append(gatewayNextHops, defaultGatewayNextHop)
+					needIPv6NextHop = false
+				}
+			}
+			if needIPv4NextHop || needIPv6NextHop {
+				return fmt.Errorf("failed to get next-hop: IPv4=%v IPv6=%v", needIPv4NextHop, needIPv6NextHop)
+			}
```
Comment (danwinship) [inline go-controller/pkg/node/gateway_init.go:166]: this is all a little bit complicated and maybe could be split out into a helper

File: go-controller/pkg/ovn/gateway_init.go [hunk 18]
```diff
@@ -316,91 +316,132 @@ func addDistributedGWPort() error {
...
+	var ipModes []string
+	if config.IPv4Mode {
+		ipModes = append(ipModes, "IPv4Mode")
+	}
+	if config.IPv6Mode {
+		ipModes = append(ipModes, "IPv6Mode")
+	}
+
+	var macStr string
+	for _, ipMode := range ipModes {
```
Comment (danwinship) [inline go-controller/pkg/ovn/gateway_init.go:328]: hm... this is weird. I think I suggested using a loop before but I was thinking you could loop over something that already existed (`subnets` or `ips` or something).

If not maybe it would be clearer to split the per-family setup code out into a separate function and then do

    if config.IPv4Mode {
            err := setupOneWhatever(util.V4NodeLocalDistributedGwPortIP, util.V4NodeLocalNatSubnetNextHop)
            if err != nil { ... }
    }
    if config.IPv6Mode {
            err := setupOneWhatever(util.V6NodeLocalDistributedGwPortIP, util.V6NodeLocalNatSubnetNextHop)
            if err != nil { ... }
    }

File: go-controller/pkg/ovn/gateway_test.go [hunk 19]
```diff
@@ -130,9 +130,9 @@ node4 chassis=912d592c-904c-40cd-9ef1-c2e5b49a33dd lb_force_snat_ip=100.64.0.4`,
 			"ovn-nbctl --timeout=15 -- --may-exist lr-add GR_test-node -- set logical_router GR_test-node options:chassis=SYSTEM-ID external_ids:physical_ip=fd99::2 external_ids:physical_ips=fd99::2",
 			"ovn-nbctl --timeout=15 -- --may-exist ls-add join_test-node",
 			"ovn-nbctl --timeout=15 -- --may-exist lsp-add join_test-node jtor-GR_test-node -- set logical_switch_port jtor-GR_test-node type=router options:router-port=rtoj-GR_test-node addresses=router",
-			"ovn-nbctl --timeout=15 -- --if-exists lrp-del rtoj-GR_test-node -- lrp-add GR_test-node rtoj-GR_test-node 0a:58:fd:98:00:01 fd98::1/125",
+			"ovn-nbctl --timeout=15 -- --if-exists lrp-del rtoj-GR_test-node -- lrp-add GR_test-node rtoj-GR_test-node " + util.IPAddrToHWAddr(net.ParseIP("fd98::1")).String() + " fd98::1/125",
```
Comment (danwinship) [inline go-controller/pkg/ovn/gateway_test.go:133]: you can hardcode things in the test code...

File: go-controller/pkg/ovn/loadbalancer.go [hunk 20]
```diff
@@ -188,14 +188,28 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb string, sourceIP string, s
 	vip := util.JoinHostPortInt32(sourceIP, sourcePort)
 	// NOTE: doesn't use vip, to avoid having brackets in the name with IPv6
 	aclName := fmt.Sprintf("%s-%s:%d", lb, sourceIP, sourcePort)
+	aclName = strings.ReplaceAll(aclName, ":", "\\:")
+	// ACL names are limited to 63 characters
+	if len(aclName) > 63 {
```
Comment (danwinship) [inline go-controller/pkg/ovn/loadbalancer.go:193]: This should be a separate commit, if it's not related to shared-gateway IPv6 support. But also, if this is a problem we're hitting, we should probably just change how we name things, rather than having a complicated truncate-in-some-cases system

File: go-controller/pkg/ovn/master.go [hunk 21]
```diff
@@ -449,14 +458,20 @@ func (oc *Controller) syncGatewayLogicalNetwork(node *kapi.Node, l3GatewayConfig
...
 		}
-		mpIP := util.GetNodeManagementIfAddr(subnets[0]).IP.String()
-
-		if err := addPolicyBasedRoutes(node.Name, mpIP, l3GatewayConfig.IPAddresses); err != nil {
-			return err
-		}
+		for _, subnet := range subnets {
+			hostIfAddr := util.GetNodeManagementIfAddr(subnet)
+			for _, l3GatewayConfigIP := range l3GatewayConfig.IPAddresses {
+				if utilnet.IsIPv6(hostIfAddr.IP) == utilnet.IsIPv6(l3GatewayConfigIP.IP) {
```
Comment (danwinship) [inline go-controller/pkg/ovn/master.go:464]: Rather than having 2 nested loops, you should be able to loop over only `subnets`, and use `util.MatchIPFamily` to find the `l3GatewayConfig.IPAddresses` value that goes with each subnet.

File: go-controller/pkg/ovn/gateway_init.go [hunk 22]
```diff
@@ -425,31 +466,47 @@ func addPolicyBasedRoutes(nodeName, mgmtPortIP string, hostIfAddrs []*net.IPNet)
...
+		externalIPs = append(externalIPs, externalIP)
 		defer func() {
 			// Release the allocation on error
 			if err != nil {
-				_ = oc.nodeLocalNatIPAllocator.Release(externalIP)
+				if isIPv6 {
+					_ = oc.nodeLocalNatIPv6Allocator.Release(externalIP)
+				} else {
+					_ = oc.nodeLocalNatIPv4Allocator.Release(externalIP)
+				}
```
Comment (danwinship) [inline go-controller/pkg/ovn/gateway_init.go:502]: (Note for later: the ipallocator stuff is awkward and we should make it more natively dual-stack friendly like subnetallocator.)

File: go-controller/pkg/util/nicstobridge.go [hunk 23]
```diff
@@ -218,11 +219,17 @@ func NicToBridge(iface string) (string, error) {
 	setupDefaultFile()
 
 	// Get ip addresses and routes before any real operations.
-	addrs, err := netlink.AddrList(ifaceLink, syscall.AF_INET)
+	family := syscall.AF_INET
+	if config.IPv6Mode && !config.IPv4Mode {
+		family = syscall.AF_INET6
+	} else if config.IPv6Mode && config.IPv4Mode {
+		family = syscall.AF_UNSPEC
```
Comment (danwinship) [inline go-controller/pkg/util/nicstobridge.go:226]: Actually, we should just use `AF_UNSPEC` here regardless of config. We are moving the entire interface onto the bridge. Even if ovn-kubernetes is running IPv4-only, if `eth0` has an IPv6 address, we should move that IPv6 address onto the bridge, or else it becomes inaccessible.

(Likewise in BridgeToNic)

**Comment by Billy99** [comment]
> Is any of this non-WIP? We can split out and merge any probably-right-and-definitely-at-least-not-breaking-IPv4 parts.

@danwinship 
I reverted back to the default upstream CI test cases (removed the temporary test.yml changes I made to limit and speed up CI)  and all passed. So to answer your question, IPv4 looks to be OK if you want all of this in. I pulled two of the smaller features out into separate commits within this PR, but there is still one big commit that I was spending too much time on trying to break up (sorry, I'll do better next time).

**Please note: IPv6 still has issues and I am continuing to debug.**

File: go-controller/pkg/ovn/gateway_test.go (hunk 19 above)
Comment (danwinship) [reply go-controller/pkg/ovn/gateway_test.go:133]: yeah, in general everything is hardcoded. let's stick with that

File: go-controller/pkg/util/nicstobridge.go [hunk 24]
```diff
@@ -218,11 +219,16 @@ func NicToBridge(iface string) (string, error) {
 	setupDefaultFile()
 
 	// Get ip addresses and routes before any real operations.
-	addrs, err := netLinkOps.AddrList(ifaceLink, syscall.AF_INET)
+	family := syscall.AF_UNSPEC
+	// FIXME: Should move all interfaces over, but in IPv4 only, breaks.
```
Comment (danwinship) [inline go-controller/pkg/util/nicstobridge.go:223]: how does it break?

oh, you probably only want to move the `.IsGlobalUnicast()` IPs over... (ie, in particular you don't want to try to move IPv6 link-local addresses)

File: go-controller/pkg/node/gateway_init.go [hunk 25]
```diff
@@ -82,15 +82,79 @@ func getIPv4Address(iface string) (*net.IPNet, error) {
...
+		needIPv4NextHop = true
+	}
+	if config.IPv6Mode {
+		needIPv6NextHop = true
+	}
+
+	// FIXME DUAL-STACK: config.Gateway.NextHop should be a slice of nexthops
+	if config.Gateway.NextHop != "" {
+		if config.IPv4Mode && !utilnet.IsIPv6String(config.Gateway.NextHop) {
+			gatewayNextHops = append(gatewayNextHops, net.ParseIP(config.Gateway.NextHop))
```
Comment (danwinship) [inline go-controller/pkg/node/gateway_init.go:125]: should probably parse `config.Gateway.NextHop` first (after checking it's not `""`), and return an error if it doesn't parse. (The current code doesn't do this; it ends up treating an invalid NextHop the same as not specifying NextHop.)

File: go-controller/pkg/node/helper_linux.go [hunk 26]
```diff
@@ -15,27 +16,62 @@ import (
...
+		family = syscall.AF_INET
+		needIPv4 = true
+	}
+	if config.IPv6Mode {
+		family = syscall.AF_INET6
+		needIPv6 = true
+	}
+	if config.IPv4Mode && config.IPv6Mode {
+		family = syscall.AF_UNSPEC
+	}
```
Comment (danwinship) [inline go-controller/pkg/node/helper_linux.go:37]: You can always use `AF_UNSPEC`; the wrong-family addresses will just get ignored anyway because, eg, `needIPv6` isn't set. So all this is just

    needIPv4 := config.IPv4Mode
    needIPv6 := config.IPv6Mode

File: go-controller/pkg/ovn/gateway_init.go [hunk 27]
```diff
@@ -316,12 +316,33 @@ func addDistributedGWPort() error {
...
+		dgpIP := net.ParseIP(util.V6NodeLocalDistributedGwPortIP)
+		// If IPv6 Only, mac won't be set. generate mac and add lrp-add cmd.
+		if dgpMac == "" {
+			dgpMac = util.IPAddrToHWAddr(dgpIP).String()
+			nbctlArgs = append(nbctlArgs,
+				"--may-exist", "lrp-add", ovnClusterRouter, dgpName, dgpMac,
+				fmt.Sprintf("%s/%d", dgpIP.String(), util.V6NodeLocalNatSubnetPrefix),
+			)
+		} else {
+			//Dual-stack. Keep same mac and append IPv6 Addess
```
Comment (danwinship) [inline go-controller/pkg/ovn/gateway_init.go:341]: space after `//`, "AddRess"

File: go-controller/pkg/ovn/gateway_init.go [hunk 28]
```diff
@@ -316,12 +316,33 @@ func addDistributedGWPort() error {
...
+			nbctlArgs = append(nbctlArgs,
+				fmt.Sprintf("%s/%d", dgpIP.String(), util.V6NodeLocalNatSubnetPrefix),
```
Comment (danwinship) [inline go-controller/pkg/ovn/gateway_init.go:343]: I don't like how this non-obivously assumes what the ipv4 branch left in `nbctlArgs`. It would be better to add the "constant" part of the args first, then have ipv4/ipv6/dual-stack parts to add the ipv4/ipv6/dual-stack addresses

File: go-controller/pkg/util/nicstobridge.go (hunk 24 above)
Comment (danwinship) [reply go-controller/pkg/util/nicstobridge.go:223]: (This could happen later though if you want to get the current version in now)

File: go-controller/pkg/ovn/gateway_init.go (hunk 22 above)
Comment (Billy99) [reply go-controller/pkg/ovn/gateway_init.go:510]: Done - added a note in ovn.go where nodeLocalNatIPv4Allocator & nodeLocalNatIPv6Allocator are defined.

File: go-controller/pkg/util/nicstobridge.go (hunk 24 above)
Comment (Billy99) [reply go-controller/pkg/util/nicstobridge.go:223]: Added the `.isGlobalUnicast()`, but still need this hack.

File: go-controller/pkg/ovn/gateway_test.go (hunk 19 above)
Comment (Billy99) [reply go-controller/pkg/ovn/gateway_test.go:133]: Done

File: go-controller/pkg/util/nicstobridge.go (hunk 24 above)
Comment (trozet) [reply go-controller/pkg/util/nicstobridge.go:227]: dont worry too much about this. I think only KIND actually moves the interface. OpenShift doesn't and neither does Nvidia. So whatever hack you are doing here is OK for now.

File: go-controller/pkg/node/gateway_init.go [hunk 29]
```diff
@@ -82,15 +82,84 @@ func getIPv4Address(iface string) (*net.IPNet, error) {
...
+			gatewayNextHops = append(gatewayNextHops, nextHop)
+			needIPv4NextHop = false
+		}
+		if config.IPv6Mode && utilnet.IsIPv6(nextHop) {
+			gatewayNextHops = append(gatewayNextHops, nextHop)
+			needIPv6NextHop = false
+		}
+	}
+	gatewayIntf := config.Gateway.Interface
+	if needIPv4NextHop || needIPv6NextHop || gatewayIntf == "" {
```
Comment (trozet) [inline go-controller/pkg/node/gateway_init.go:139]: this if statement doesn't make sense to me. if you need ipv4 or ipv6 next hops, but gatewayIntf is provided, you shouldn't just do getDefaultGatewayInterfaceDetails right? That will search all routes instead of just the ones for the gateway interface. It seems like it should be:

```
if needIPv4NextHop || needIPv6NextHop {
         if gatewayIntf == "" {
              defaultGatewayIntf, defaultGatewayNextHops, err := getDefaultGatewayInterfaceDetails()
         } else {
            // find ipv4 and ipv6 routes for gatewayIntf
        }

```

**Comment by trozet** [comment]
@Billy99 can you run this downstream and see if ipv6 job works?

**Comment by trozet** [comment]
@Billy99 oh also, add another commit here to make the local ipv6 job shared gw ipv6

File: go-controller/pkg/node/gateway_init.go (hunk 29 above)
Comment (danwinship) [reply go-controller/pkg/node/gateway_init.go:139]: This matches the older single-stack logic. The autodetection can only handle simple cases. Of course, given that you can only manually specify a single NextHop value right now, this means you can't currently do dual stack shared gateway mode on a node with multiple default gateways where the one it autodetects isn't the one you wanted. To be revisited.

**Comment by danwinship** [comment]
ok, this at least improves things as-is. More work to be done...

**Comment by trozet** [comment]
@danwinship @Billy99 let's see how it goes: #1613
